floris/wind_data.py

```diff
@@ -2651,3 +2672,672 @@ def to_WindTIRose(
+        if (
+            x < self.x_array[0]
+            or x > self.x_array[-1]
+            or y < self.y_array[0]
+            or y > self.y_array[-1]
+        ):
+            method = "nearest"
+        else:
+            method = "linear"
+
+        result = np.zeros(self.n_sectors)
+        for sector in range(self.n_sectors):
+            result[sector] = interpolant_function_list[sector]((x, y), method=method)
+
+        return result
+
+    def _weibull_cumulative(self, x, a, k):
+        """
+        Calculate the Weibull cumulative distribution function.
+
+        Args:
+            x (np.array): The wind speed values.
+            a (np.array): The Weibull A parameter values.
+            k (np.array): The Weibull k parameter values.
+
+        Returns:
+            np.array: The cumulative distribution function values.
+        """
+
+        exponent = -((x / a) ** k)
+        result = 1.0 - np.exp(exponent)
+
+        # Where x is less than 0, the result should be 0
+        result[x < 0] = 0.0
+
+        return result
+
+        # Original code from PJ Stanley
+        # if x >= 0.0:
+        #     exponent = -(x / a) ** k
+        #     return 1.0 - np.exp(exponent)
+        # else:
+        #     return 0.0
+
+    def _generate_wind_speed_frequencies_from_weibull(self, A, k, wind_speeds=None):
+        """
+        Generate the wind speed frequencies from the Weibull parameters.  Use the
+        cumulative form of the function and calculate the probability of the wind speed
+        in a given bin via the difference in the cumulative function at the bin edges.
+        Args:
+
+            A (np.array): The Weibull A parameter.
+            k (np.array): The Weibull k parameter.
+            wind_speeds (np.array): The wind speeds to calculate the frequencies for.
+                If None, the frequencies are calculated for 0 to 25 m/s in 1 m/s increments.
+                Default is None.
+
+        Returns:
+            np.array: The wind speed frequencies.
+        """
+
+        if wind_speeds is None:
+            wind_speeds = self.wind_speeds
+
+        # Define the wind speed edges
+        ws_step = wind_speeds[1] - wind_speeds[0]
+        wind_speed_edges = np.arange(
+            wind_speeds[0] - ws_step / 2, wind_speeds[-1] + ws_step, ws_step
+        )
+
+        # Get the cumulative distribution function at the edges
+        cdf_edges = self._weibull_cumulative(wind_speed_edges, A, k)
+
+        # The frequency is the difference in the cumulative distribution function
+        # at the edges
+        freq = cdf_edges[1:] - cdf_edges[:-1]
+
+        # Normalize the frequency
+        # TODO: This is perhaps not quite right, if the user only asks
+        # for wind speeds at say 8 and 10 m/s
+        # then the bins would be from 7 to 11 and so fairly there is
+        # distribution outside the range, but
+        # I think this is on the whole right -- PF
```

Comment:
Does this need to be addressed still?

Reply:
I guess I think now it's right enough.  When I wrote this comment I was thinking renormalzing when we know certain bands are not included is wrong, but now I think it's just another interpretation.  Like, the Weibull distribution in theory goes on forever but it seems ok to say that we want the bins which we sampled to sum to 1.0.  I think that changes slightly the meaning but in a way which FLORIS expects, we don't really have a mechanism for wind speed bin weights that don't sum to 1.0.   So I think we could delete this TODO if you agree @misi9170

Reply:
I think the current approach is correct, if I'm understanding everything correctly. However, I do think we could add a more detailed comment to avoid any confusion in future, something along the lines of:
```
# NOTE: The probability mass associated to each discrete wind speed (ws) is taken as the cumulative mass under the continuous Weibull distribution from ws-ws_step to ws+ws_step, where ws_step is the step between the provided wind_speeds.
```

In looking more carefully at this,

Reply:
I think enforcing even step sizes is good approach for now, and matches the requirements put in for wind direction in the upsample pull request.  I think even spacing is implicitly assumed so good to made it explicit and if we decide to relax it to allow un-even spacing we need to handle/test a number of cases, if there's a good reason to support it

Reply:
Done.